lib/render/ParticleRenderer.cpp

```diff
@@ -178,4 +190,217 @@ int ParticleRenderer::_paintGL(bool)
     return rc;
 }
 
-int ParticleRenderer::_initializeGL() { return 0; }
+int ParticleRenderer::_initializeGL() { 
+    glGenVertexArrays(1, &_vertexArrayId);
+    glGenBuffers(1, &_vertexBufferId);
+
+    /* Generate and configure 1D texture: _colorMapTexId */
+    glGenTextures(1, &_colorMapTexId);
+    glActiveTexture(GL_TEXTURE0 + _colorMapTexOffset);
+    glBindTexture(GL_TEXTURE_1D, _colorMapTexId);
+    glTexParameteri(GL_TEXTURE_1D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+    glTexParameteri(GL_TEXTURE_1D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+    glTexParameteri(GL_TEXTURE_1D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+    glBindTexture(GL_TEXTURE_1D, 0);
+
+    glGenVertexArrays(1, &_VAO);
+    glGenBuffers(1, &_VBO);
+
+    assert(_VAO);
+    assert(_VBO);
+
+    glBindVertexArray(_VAO);
+    glBindBuffer(GL_ARRAY_BUFFER, _VBO);
+    glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, sizeof(glm::vec4), NULL);
+    glVertexAttribPointer(1, 1, GL_FLOAT, GL_FALSE, sizeof(glm::vec4), (void *)sizeof(glm::vec3));
+    glEnableVertexAttribArray(0);
+    glEnableVertexAttribArray(1);
+    glBindBuffer(GL_ARRAY_BUFFER, 0);
+    glBindVertexArray(0);
+    return 0; 
+}
+
+int ParticleRenderer::_renderParticles(const std::vector<glm::vec4>& particles)
+{
+    typedef struct {
+        glm::vec3  p;
+        float v;
+    } Vertex;
+    vector<Vertex> vertices;
+    vector<int>    sizes;
+    vector<Vertex> sv;
+
+    int nStreams = particles.size();
+    for (int s = 0; s < nStreams; s+=2) {
+        glm::vec4 p = particles[s];
+        glm::vec4 p2 = particles[s+1];
+        const vector<flow::Particle> stream = {flow::Particle(p[0],p[1],p[2],0.,p[3]), flow::Particle(p2[0],p2[1],p2[2],0.,p2[3])};
```

Comment:
Do we want to convert each particle to a flow "Stream"? At the very least if we are doing this why copy and paste all the same code?

Reply:
> Do we want to convert each particle to a flow "Stream"?

Doing it this way made it convenient to draw the particle directions.

> At the very least if we are doing this why copy and paste all the same code?

There are several differences between ParticleRenderer::_renderParticles() and FlowRenderer::_renderStreamlines(), but many of the lines are copied.  As to why I did it this way, it wasn't just being haphazard.  

There' a rule of thumb called [the rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) that suggests refactoring take place after code has been duplicated three times.  As a rule of thumb, this is open for debate.  But if we do choose to refactor both this and the FlowRenderer, we need to agree on exactly how it should be done.

Reply:
In the recent revision, I've removed the usage of flow::Particle.  There are still a handfull of lines in ParticleRenderer.cpp::_renderParticleGlyphs() that match the FlowRenderer.cpp::_renderAdvection(), but not enough commonality to share logic between the two classes, IMO.